torch/csrc/jit/passes/onnx/shape_type_inference.cpp

```diff
@@ -456,9 +459,133 @@ void ONNXAssignOutputShape(
+// output of the other subblock. prim::Uninitialized node is proven to be
+// unused. So replace this node with a constant of the inferred shape and type.
+void InferShapeTypeForUninitializedOutput(
+    Graph* graph,
+    Block* block,
+    Value* uninitialized_output,
+    Value* other_output) {
+  auto output_type = other_output->type()->expect<TensorType>();
+  auto elem_type = at::initialTensorOptions().dtype(output_type->scalarType());
+  Node* const_node = graph->create(::c10::onnx::Constant, 1);
+
+  if (output_type->sizes().concrete_sizes().has_value()) {
+    auto size = output_type->sizes().concrete_sizes().value();
+    const_node->t_(attr::value, at::zeros(size, elem_type));
+    const_node->output()->setType(other_output->type());
+  } else {
+    const_node->t_(attr::value, at::zeros({}, elem_type));
+    const_node->output()->setType(
+        TensorType::create(*(output_type->scalarType()), at::kCPU, {}, {}));
+  }
+  const_node->insertBefore(block->return_node());
+  const_node->output()->copyMetadata(other_output);
+  uninitialized_output->replaceAllUsesWith(const_node->output());
+  uninitialized_output->node()->destroy();
+}
+
+// Corresponding outputs for ONNX If then and else subblocks should have
+// same shape and type. This pass detects if prim::Uninitialized node
+// appears as part of outputs of either of the subblocks, and infers
+// shape and type from the corresponding output of the other subblock
+// In the example graph below, shape and type of the subblock output %7
+// for subblock 1 is inferred from %y.1. Shape and type of Subblock
+// output %7 is inferred from %y.5.
+//
+// graph(%y.1 : Int(3:4, 4:1, requires_grad=0, device=cpu)):
+//   ...
+//   %7 : Tensor = prim::Uninitialized()
+//   %16 : bool, %17 : Tensor, %y.14 : Tensor = prim::If(%15) #
+//   test/onnx/test_pytorch_onnx_onnxruntime.py:614:20
+//     block0():
+//       %y.5 : Tensor = aten::add(%y.1, %3, %6) #
+//       test/onnx/test_pytorch_onnx_onnxruntime.py:615:28
+//       -> (%2, %7, %y.5)
+//     block1():
+//       -> (%1, %y.1, %7)
+//   ...
+
+void ONNXIfShapeTypeInference(Node* node) {
+  for (Block* b : node->blocks()) {
+    for (Node* n : b->nodes()) {
+      if (n->kind() == ::c10::onnx::If) {
+        ONNXIfShapeTypeInference(n);
+      }
+    }
+  }
+
+  if (node->kind() != ::c10::onnx::If) {
+    return;
+  }
+
+  GRAPH_DUMP("Graph before fixing If shape type: ", node->owningGraph());
+  auto* if_node = node;
+  auto* graph = if_node->owningGraph();
+
+  // Check if the input to ONNX If node is node Bool, and insert
+  // cast to Bool if needed.
+  if (!if_node->input()->type()->isSubtypeOf(BoolType::get())) {
+    Node* cast_node = CreateCastToBoolNode(if_node->input(), graph);
+    cast_node->insertBefore(if_node);
+    if_node->replaceInputWith(if_node->input(), cast_node->output());
+  }
+
+  Block* then_block = if_node->blocks()[0];
+  Block* else_block = if_node->blocks()[1];
```

Comment:
does if block always have 2 sub-blocks then and else? Can there be multiple else sub-blocks, for example if-elif-elif-else?

Reply:
No, all onnx If nodes have 2 subblocks only.